src/common/ScreenExporter.cpp

```diff
@@ -0,0 +1,198 @@
+        ihdr[1]  = (width >> 16) & 0xFF;
+        ihdr[2]  = (width >>  8) & 0xFF;
+        ihdr[3]  = (width >>  0) & 0xFF;
+        ihdr[4]  = (height >> 24) & 0xFF;  // height
+        ihdr[5]  = (height >> 16) & 0xFF;
+        ihdr[6]  = (height >>  8) & 0xFF;
+        ihdr[7]  = (height >>  0) & 0xFF;
+        ihdr[8]  = 8;  // 8 bits per sample (24 bits per pixel)
+        ihdr[9]  = 2;  // PNG_COLOR_TYPE_RGB
+        ihdr[10] = 0;  // PNG_COMPRESSION_TYPE_DEFAULT
+        ihdr[11] = 0;  // PNG_FILTER_TYPE_DEFAULT
+        ihdr[12] = 0;  // PNG_INTERLACE_NONE
+        writePNGChunk(out, "IHDR", ihdr, sizeof(ihdr));
+}
+
+
+static void writePNGData(std::ofstream &out, const ALEScreen &screen, const ColourPalette &palette, bool doubleWidth = true) {
+
+    int dataWidth = screen.width(); 
+    int width = doubleWidth ? dataWidth * 2 : dataWidth; 
+    int height = screen.height();
+   
+    // If so desired, double the width
+
+    // Fill the buffer with scanline data
+    int rowbytes = width * 3;
+
+    std::vector<uInt8> buffer((rowbytes + 1) * height, 0);
+    uInt8* buf_ptr = &buffer[0];
+
+    for(int i = 0; i < height; i++) {
+        *buf_ptr++ = 0;                  // first byte of row is filter type
+        for(int j = 0; j < dataWidth; j++) {
+            int r, g, b;
+
+            palette.getRGB(screen.getArray()[i * dataWidth + j], r, g, b);
+            // Double the pixel width, if so desired
+            int jj = doubleWidth ? 2 * j : j;
+
+            buf_ptr[jj * 3 + 0] = r;
+            buf_ptr[jj * 3 + 1] = g;
+            buf_ptr[jj * 3 + 2] = b;
+            
+            if (doubleWidth) {
+                
+                jj = jj + 1;
+
+                buf_ptr[jj * 3 + 0] = r;
+                buf_ptr[jj * 3 + 1] = g;
+                buf_ptr[jj * 3 + 2] = b;
+            }
+        }
+        buf_ptr += rowbytes;                 // add pitch
+    }
+
+    // Compress the data with zlib
+    uLongf compmemsize = (uLongf)((height * (width + 1) * 3 + 1) + 12);
+    std::vector<uInt8> compmem(compmemsize, 0);
+    
+    if((compress(&compmem[0], &compmemsize, &buffer[0], height * (width * 3 + 1)) != Z_OK)) {
+
+        // @todo -- throw a proper exception
+        std::cerr << "Error: Couldn't compress PNG" << std::endl;
+        return;
+    }
+
+    // Write the compressed framebuffer data
+    writePNGChunk(out, "IDAT", &compmem[0], compmemsize);
+}
+
+
+static void writePNGEnd(std::ofstream &out) {
+
+    // Finish up
+    writePNGChunk(out, "IEND", 0, 0);
+}
+
+ScreenExporter::ScreenExporter(ColourPalette &palette):
+    m_palette(palette),
+    m_frame_number(0),
+    m_frame_field_width(6),
+    m_basename("") {
+}
+
+
+ScreenExporter::ScreenExporter(ColourPalette &palette, const std::string &basename):
+    m_palette(palette),
+    m_frame_number(0),
+    m_frame_field_width(6),
+    m_basename(basename) {
+
+    // MGB: Should we strip '.png' from filename if present?
```

Comment:
Sorry I just meant the png writer will fail though if a bad file path is specified (e.g. /tmp/qbert/ doesnt exist). Okay, so now after your example I see what you mean by the basename/basepath. We should document clearly (maybe with an example) so that others don't get confused like me. I agree with you that the special character (X,#) way is clunky. I'd argue that calling it basepath is more clear to me than basename.